src/test/sv2_template_provider_tests.cpp

```diff
@@ -0,0 +1,314 @@
+    BOOST_REQUIRE(tester.start());
+
+    BOOST_REQUIRE(!tester.IsConnected());
+    tester.handshake();
+    BOOST_REQUIRE(tester.IsConnected());
+    BOOST_REQUIRE(!tester.IsFullyConnected());
+
+    // After the handshake the client must send a SetupConnection message to the
+    // Template Provider.
+
+    // An empty SetupConnection message should cause disconnection
+    node::Sv2NetHeader setup_conn_header{node::Sv2MsgType::SETUP_CONNECTION, 0};
+    node::Sv2NetMsg sv2_msg{std::move(setup_conn_header), {}};
+    tester.receiveMessage(sv2_msg);
+    tester.ProcessOurResponse(0);
+
+    BOOST_REQUIRE(!tester.IsConnected());
+
+    // Reconnect
+    tester.handshake();
+
+    node::Sv2NetMsg setup{tester.SetupConnectionMsg()};
+    // SetupConnection.Success is 6 bytes
+    tester.receiveMessage(setup);
+    tester.ProcessOurResponse(SV2_HEADER_ENCRYPTED_SIZE + 6 + Poly1305::TAGLEN);
+    BOOST_REQUIRE(tester.IsFullyConnected());
+
+    // There should be no block templates before any client gave us their coinbase
+    // output data size:
+    BOOST_REQUIRE(tester.m_tp->GetBlockTemplates().empty());
+
+    std::vector<uint8_t> coinbase_output_max_additional_size_bytes{
+        0x01, 0x00, 0x00, 0x00
+    };
+    node::Sv2NetHeader cb_header{node::Sv2MsgType::COINBASE_OUTPUT_DATA_SIZE, 4};
+    node::Sv2NetMsg msg{std::move(cb_header), std::move(coinbase_output_max_additional_size_bytes)};
+    // The reply should be NewTemplate and SetNewPrevHash
+    tester.receiveMessage(msg);
+    tester.ProcessOurResponse(2 * SV2_HEADER_ENCRYPTED_SIZE + 91 + 80 + 2 * Poly1305::TAGLEN);
+
+    // There should now be one template
+    BOOST_REQUIRE_EQUAL(tester.m_tp->GetBlockTemplates().size(), 1);
+
+    // Move mock time by at least DEFAULT_SV2_INTERVAL
+    // If the mempool doesn't change, no new template is generated.
+    SetMockTime(GetMockTime() + std::chrono::seconds{DEFAULT_SV2_INTERVAL});
+    // Briefly wait for the timer in ThreadSv2Handler and block creation
+    UninterruptibleSleep(std::chrono::milliseconds{200});
+    BOOST_REQUIRE_EQUAL(tester.m_tp->GetBlockTemplates().size(), 1);
+
+    // Create a transaction with a large fee
+    size_t tx_size;
+    CKey key = GenerateRandomKey();
+    CScript locking_script = GetScriptForDestination(PKHash(key.GetPubKey()));
+    // Don't hold on to the transaction
+    {
+        auto mtx = CreateValidMempoolTransaction(/*input_transaction=*/m_coinbase_txns[0], /*input_vout=*/0,
+                                                        /*input_height=*/0, /*input_signing_key=*/coinbaseKey,
+                                                        /*output_destination=*/locking_script,
+                                                        /*output_amount=*/CAmount(49 * COIN), /*submit=*/true);
+        CTransactionRef tx = MakeTransactionRef(mtx);
+
+        // Get serialized transaction size
+        DataStream ss;
+        ss << TX_WITH_WITNESS(tx);
+        tx_size = ss.size();
+    }
+
```

Comment:
ASAN is tripping up somewhere around here. The last thing it logs is ` - Connect 2 transactions:`. It doesn't get to ` - Verify ... txins:`. I wonder if this is related to mock time, which I'm testing in https://github.com/Sjors/bitcoin/pull/34

Reply:
@maflcko shouldn't TSan on CI output something useful about why it crashed? I currently only says "error 2": https://cirrus-ci.com/task/5124733717446656?logs=ci#L3531

When running this locally on Ubuntu with clang 16.0.6 I get a `WARNING: ThreadSanitizer: data race` and significantly more details (still a bit cryptic, but hopefully enough to figure out what's happening).